Cpp_test/lib/ipc-shm-send.cpp

```diff
@@ -0,0 +1,114 @@
+#include "ipc-shm-send.h"
+#include <iostream>
+#include <cassert>
+
+IPCShmSend::~IPCShmSend()
+{
+
+	shm_unlink(this->info.method_name);
+	
+}
+
+void IPCShmSend::init()
+{
+	this->file_handler.open_file();
+	this->open_shm();
+	this->set_shm_size();
+	this->map_shm();
+	if (this->shmd > 0)
+	{
+		close(this->shmd);
+	}
+	this->init_mutex();
+	this->init_cond();
+}
+
+void IPCShmSend::transfer()
+{
+	long read_bytes{0};
+	unsigned long long file_size{this->file_handler.get_file_size()};
+	if (file_size == 0)
+	{
+		throw std::runtime_error("ERROR: File size = 0.");
+	}
+
+	unsigned long long total_sent_bytes{0};
+	std::cout << "Waiting for receiver..." << std::endl;
+	while (!this->is_end)
+	{
+		lock_mutex();
+		while (this->shm_ptr->data_version_received != this->shm_ptr->data_version)
+		{
+
+			if (pthread_cond_wait(&(this->shm_ptr->cond_re), &(this->shm_ptr->mutex)) != 0)
+			{
+				throw std::runtime_error(std::string("ERROR: pthread_cond_wait() send: ") + strerror(errno));
+			}
+		}
+
+		if (total_sent_bytes < file_size)
+		{
+			this->file_handler.read_file_shm(this->shm_ptr->data_ap, this->shm_ptr->data_size);
+			read_bytes = this->file_handler.get_read_bytes();
+
+			if (read_bytes > 0)
+			{
+				this->shm_ptr->data_version++;
+				this->shm_ptr->data_size = read_bytes;
+				total_sent_bytes += read_bytes;
+			}
+		}
+		else
+		{
+			this->shm_ptr->data_size = 0;
+			this->shm_ptr->data_version++;
+			this->is_end = true;
+		}
+		send_cond_broadcast();
+		unlock_mutex();
+	}
+
+	if (total_sent_bytes == file_size && this->is_end)
+	{
+		std::cout << "Sent: " << total_sent_bytes << " byte(s)." << std::endl;
+	}
+	else
+	{
+		throw std::runtime_error("ERROR: The size of Total send file is not equal to File size.");
+	}
+	pthread_mutexattr_destroy(&(this->mutex_attr));
+}
+
+void IPCShmSend::map_shm()
+{
+	this->shm_ptr = (ipc_shm_header_t *)mmap64(NULL, sizeof(ipc_shm_header_t),
+											   PROT_READ | PROT_WRITE, MAP_SHARED,
+											   this->shmd, 0);
+	if ((static_cast<void *>(this->shm_ptr)) == MAP_FAILED)
+	{
+		throw std::runtime_error("ERROR: mmap64().");
+	}
+	this->shm_ptr->data_ap = (char *)mmap64(NULL, 2048,
+											PROT_READ | PROT_WRITE, MAP_SHARED,
+											this->shmd, 4096);
+	if ((static_cast<void *>(this->shm_ptr->data_ap)) == MAP_FAILED)
+	{
+		throw std::runtime_error("ERROR: mmap64() data_ap.");
+	}
+	// set values for new share memory
+	this->shm_ptr->is_init = true;
+	this->shm_ptr->shared_mem_size = this->shm_size_in_bytes;
+	this->shm_ptr->data_size = 2048;
+}
+
+void IPCShmSend::open_shm()
+{
+	// remove old shared memory (if any)
+	shm_unlink(this->info.method_name);
+	errno = 0;
```

Comment:
errno should not be changed manually.

(this comment is valid for all the program)